### PR TITLE
Merge `Sniffer` with `Examiner` and refactor `Examiner`.

### PR DESCRIPTION
**Pull request by troessner**
wip please dont review.

### CODE REVIEW

**Comment by troessner** [comment]
Ready for review!
This is part of #395 and it:

1.) Gets rid of the unnecessary and confusing Sniffer abstraction by merging it into Examiner
2.) Refactors Examiner.

**Comment by troessner** [comment]
@mvz would be great if you could review this bad boy before the weekend starts since I'd like to start with refactoring TreeWalker and i'd like to build on the changes in this PR.:)

**Comment by mvz** [comment]
Yeah, I should be able to review it before then.

**Comment by troessner** [comment]
@mvz any updates on this?..:)

**Comment by troessner** [comment]
@mvz No problem if you don't have the time to do this right now, but since this is blocking me from further refactorings i'll merge this one tonight if no other contributor looks into reviewing this (especially since this has been up for more than 2 weeks now).

**Comment by mvz** [comment]
Sorry, I've been meaning to look at this and not getting 'round to it. The worst combination ...

File: lib/reek/core/smell_repository.rb [hunk 1]
```diff
@@ -13,12 +13,17 @@ def self.smell_types
         Reek::Smells::SmellDetector.descendants
       end
 
+      def self.default
```
Comment (mvz) [inline lib/reek/core/smell_repository.rb]: This may be a little confusing since in the Singleton mixin this would always return the same instance.

Comment (mvz) [inline lib/reek/core/smell_repository.rb]: How about just making the first param to new nil by default and using that?

**Comment by mvz** [comment]
One tiny nit, otherwise good to merge.

**Comment by troessner** [comment]
@mvz I removed the SmellRepository.default method and made the first parameter to SmellRepositorys "new" nil by default - merge?..:)

**Comment by mvz** [comment]
@troessner excellent. I'll merge once Travis is done.

**Pull request merged**